### PR TITLE
Digitizer HID interface : absolute coordinates for mouse cursor

### PR DESCRIPTION
**Pull request by a-chol**
Digitizer HID interface : absolute coordinates for mouse cursor

## Description

I've added a HID digitizer report enabled with DIGITIZER_ENABLE, which allows positioning the mouse cursor in absolute normalized screen coordinates.
This has been implemented for LUFA and Chibios backends, and tested on the pro micro as well as a blackpill and a proton c.

A few feature based on this could follow, such as mapping key coordinates to set the cursor position or a dichotomic positioning algorithm. Users can build mouse-based macros with this.

A handwired/onekey digitizer keymap has been added that rotates the cursor around the center of the screen.

## Types of Changes

- [X] Core
- [ ] Bugfix
- [X] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [X] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

I'll get around to check codestyle if needed in the coming days.

### CODE REVIEW

**Comment by tzarc** [comment]
> I'll get around to check codestyle if needed in the coming days.

It's needed, along with all the copyright headers for new files, too.

**Comment by drashna** [comment]
Also, report information and HID endpoints need to be addded: 
https://github.com/qmk/qmk_firmware/blob/3f854e16acc880b9c9ccab2244dc585705dfac1e/tmk_core/protocol/usb_descriptor.h

And V-USB support should be added.

**Comment by a-chol** [comment]
> 
> 
> Also, report information and HID endpoints need to be addded:
> https://github.com/qmk/qmk_firmware/blob/3f854e16acc880b9c9ccab2244dc585705dfac1e/tmk_core/protocol/usb_descriptor.h
> 
> And V-USB support should be added.

I was piggybacking on the shared report for the digitizer report descriptor as well as for the endpoint. Should I had a full dedicated endpoint for digitizer with the option to DIGITIZER_SHARED_EP to use the shared endpoint?

Regarding V-USB, it seems only the shared endpoint is an option. The joystick report was removed there pending some reorg of the endpoint management in V-USB  ([4226](https://github.com/qmk/qmk_firmware/issues/4226#issuecomment-600898689_)), and I'm not sure if this has happened yet. Do I had the digitizer reporting to the shared report of V-USB?

File: docs/feature_digitizer.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+To enable the digitizer interface, add the following line to your rules.mk: 
+
+```makefile
+DIGITIZER_ENABLE = yes
+```
+
+In order to change the mouse cursor position from your keymap.c file, include the digitizer header : 
+
+```c
+#include <digitizer.h>
```
Comment (fauxpark) [inline docs/feature_digitizer.md:14]:
```suggestion
#include "digitizer.h"
```

File: keyboards/handwired/onekey/keymaps/digitizer/config.h [hunk 2]
```diff
@@ -0,0 +1,2 @@
+#pragma once
```
Comment (fauxpark) [inline keyboards/handwired/onekey/keymaps/digitizer/config.h:1]: Can be deleted.

File: keyboards/handwired/onekey/keymaps/digitizer/keymap.c [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    LAYOUT_ortho_1x1(JS_BUTTON0)
+};
+
+void matrix_scan_user() {
+    if (((timer_read()/10) % 10) != 0){
+      return;
+    }
+    digitizer.x = 0.5-0.2*cos(timer_read()/250./6.28);
+    digitizer.y = 0.5-0.2*sin(timer_read()/250./6.28);
```
Comment (fauxpark) [inline keyboards/handwired/onekey/keymaps/digitizer/keymap.c:20]:
```suggestion
    if (((timer_read() / 10) % 10) != 0) {
      return;
    }

    digitizer.x = 0.5 - 0.2 * cos(timer_read() / 250. / 6.28);
    digitizer.y = 0.5 - 0.2 * sin(timer_read() / 250. / 6.28);
```

File: quantum/process_keycode/process_digitizer.c [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#include "digitizer.h"
```
Comment (fauxpark) [inline quantum/process_keycode/process_digitizer.c:1]: Not sure this should be in `process_keycode/`, tbh. But perhaps it's the folder name that's wrong...

File: tmk_core/protocol/usb_descriptor.c [hunk 5]
```diff
@@ -227,6 +227,36 @@ const USB_Descriptor_HIDReport_Datatype_t PROGMEM SharedReport[] = {
...
+        HID_RI_REPORT_SIZE(8, 0x10),
+        HID_RI_REPORT_COUNT(8, 0x01),
+        HID_RI_UNIT(8, 0x33),
+        HID_RI_UNIT_EXPONENT(8, 0x0E),
+        HID_RI_USAGE(8, 0x30),//X
+        HID_RI_INPUT(8, HID_IOF_VARIABLE),
+        HID_RI_USAGE(8, 0x31),//Y
+        HID_RI_INPUT(8, HID_IOF_VARIABLE),
+      HID_RI_END_COLLECTION(0),
+    HID_RI_END_COLLECTION(0)
```
Comment (fauxpark) [inline tmk_core/protocol/usb_descriptor.c:258]:
```suggestion

#if DIGITIZER_ENABLE
    HID_RI_USAGE_PAGE(8, 0x0D),      // Digitizers
    HID_RI_USAGE(8, 0x02),           // Pen
    HID_RI_COLLECTION(8, 0x01),      // Application
        HID_RI_REPORT_ID(8, REPORT_ID_DIGITIZER),
        HID_RI_USAGE(8, 0x20),       // Stylus
        HID_RI_COLLECTION(8, 0x00),  // Physical
            // Tip Switch (1 bit)
            HID_RI_USAGE(8, 0x42),   // Tip Switch
            HID_RI_LOGICAL_MINIMUM(8, 0x00),
            HID_RI_LOGICAL_MAXIMUM(8, 0x01),
            HID_RI_REPORT_SIZE(8, 0x01),
            HID_RI_REPORT_COUNT(8, 0x01),
            HID_RI_INPUT(8, HID_IOF_VARIABLE),
            // In Range (1 bit)
            HID_RI_USAGE(8, 0x32),  // In Range
            HID_RI_INPUT(8, HID_IOF_VARIABLE),
            // Padding (6 bits)
            HID_RI_REPORT_COUNT(8, 0x06),
            HID_RI_INPUT(8, HID_IOF_CONSTANT | HID_IOF_VARIABLE),

            // X/Y Position (4 bytes)
            HID_RI_USAGE_PAGE(8, 0x01),     // Generic Desktop
            HID_RI_LOGICAL_MAXIMUM(16, 0x7FFF),
            HID_RI_REPORT_SIZE(8, 0x10),
            HID_RI_REPORT_COUNT(8, 0x01),
            HID_RI_UNIT(8, 0x33),           // Inch, English Linear
            HID_RI_UNIT_EXPONENT(8, 0x0E),  // -2
            HID_RI_USAGE(8, 0x30),          // X
            HID_RI_INPUT(8, HID_IOF_VARIABLE),
            HID_RI_USAGE(8, 0x31),          // Y
            HID_RI_INPUT(8, HID_IOF_VARIABLE),
        HID_RI_END_COLLECTION(0),
    HID_RI_END_COLLECTION(0)
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: tmk_core/common/host_driver.h [hunk 6]
```diff
@@ -29,4 +29,5 @@ typedef struct {
     void (*send_mouse)(report_mouse_t *);
     void (*send_system)(uint16_t);
     void (*send_consumer)(uint16_t);
+    void (*send_digitizer)(report_digitizer_t*);
```
Comment (fauxpark) [inline tmk_core/common/host_driver.h:32]: Is there a particular reason for this?

Comment (a-chol) [reply tmk_core/common/host_driver.h:32]: I was trying to stay consistent with the existing way of separating the hw-independent way of building of the report from the hw-dependent way of sending it. Do you have a suggestion for a better way to do this? I know the joystick sending is not made that way, but I think it's a mistake to mix ways of doing the same thing. And I'm allowed to criticize, I made said mistake! :D

File: keyboards/handwired/onekey/keymaps/digitizer/keymap.c (hunk 3 above)
Comment (a-chol) [reply keyboards/handwired/onekey/keymaps/digitizer/keymap.c:20]: I've changed the if to have a more understandable way of pacing the update using timer_elapsed, but I'll apply your formatting on the next commit.

File: quantum/process_keycode/process_digitizer.c (hunk 4 above)
Comment (a-chol) [reply quantum/process_keycode/process_digitizer.c:1]: Yeah, I think I got mistaken by taking example on the midi implementation that has the task function in the process_keycode folder, but other features have them in the implementation file in the parent directory. I'll fix that for the digitizer and I should make the joystick implementation more consistent with the rest in a separate PR.

**Comment by a-chol** [comment]
So after a bit of a struggle I managed to get dedicated and shared endpoints working on lufa and chibios, and implemented the V-USB implementation.
I'll check code style all around for the coming commit.

**Comment by a-chol** [comment]
hmm, if I run qmk cformat on my branch, a lot of changes on files I didn't touch appear. This is not expected. Should I keep only changes on my files?

**Comment by drashna** [comment]
> hmm, if I run qmk cformat on my branch, a lot of changes on files I didn't touch appear. This is not expected. Should I keep only changes on my files?

Yeah, just keep the changes on files you touched.

**Comment by fauxpark** [comment]
> 
> 
> hmm, if I run qmk cformat on my branch, a lot of changes on files I didn't touch appear. This is not expected. Should I keep only changes on my files?

Use the Docker container to run clang-format, different versions seem to have different ideas of what "formatted" means.

File: tmk_core/protocol/vusb/vusb.c [hunk 7]
```diff
@@ -509,8 +518,40 @@ const PROGMEM uchar shared_hid_report[] = {
...
+    0x75, 0x10,                 //     REPORT_SIZE (16)
+    0x95, 0x01,                 //     REPORT_COUNT (1)
+    0x65, 0x33,                 //     UNIT (Eng Lin:Distance)
+    0x55, 0x0e,                 //     UNIT_EXPONENT (-2)
+    0x09, 0x30,                 //     USAGE (X)
+    0x81, 0x02,                 //     INPUT (Data,Var,Abs)
+    0x09, 0x31,                 //     USAGE (Y)
+    0x81, 0x02,                 //     INPUT (Data,Var,Abs)
+    0xc0,                       //   END_COLLECTION
+    0xc0,                       // END_COLLECTION
```
Comment (fauxpark) [inline tmk_core/protocol/vusb/vusb.c:552]:
```suggestion
    // Digitizer report descriptor
    0x05, 0x0D,                 // Usage Page (Digitizers)
    0x09, 0x01,                 // Usage (Digitizer)
    0xA1, 0x01,                 // Collection (Application)
    0x85, REPORT_ID_DIGITIZER,  //   Report ID
    0x09, 0x22,                 //   Usage (Finger)
    0xA1, 0x00,                 //   Collection (Physical)
    // Tip Switch (1 bit)
    0x09, 0x42,  //     Usage (Tip Switch)
    0x15, 0x00,  //     Logical Minimum
    0x25, 0x01,  //     Logical Maximum
    0x95, 0x01,  //     Report Count (1)
    0x75, 0x01,  //     Report Size (16)
    0x81, 0x02,  //     Input (Data, Variable, Absolute)
    // In Range (1 bit)
    0x09, 0x32,  //     Usage (In Range)
    0x81, 0x02,  //     Input (Data, Variable, Absolute)
    // Padding (6 bits)
    0x95, 0x06,  //     Report Count (6)
    0x81, 0x03,  //     Input (Constant)

    // X/Y Position (4 bytes)
    0x05, 0x01,        //     Usage Page (Generic Desktop)
    0x26, 0xFF, 0x7F,  //     Logical Maximum (32767)
    0x95, 0x01,        //     Report Count (1)
    0x75, 0x10,        //     Report Size (16)
    0x65, 0x33,        //     Unit (Inch, English Linear)
    0x55, 0x0E,        //     Unit Exponent (-2)
    0x09, 0x30,        //     Usage (X)
    0x81, 0x02,        //     Input (Data, Variable, Absolute)
    0x09, 0x31,        //     Usage (Y)
    0x81, 0x02,        //     Input (Data, Variable, Absolute)
    0xC0,              //   End Collection
    0xC0               // End Collection
```

File: keyboards/handwired/onekey/keymaps/digitizer/config.h [hunk 8]
```diff
@@ -0,0 +1 @@
+
```
Comment (fauxpark) [inline keyboards/handwired/onekey/keymaps/digitizer/config.h:1]: This file should be deleted.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: tmk_core/common/host_driver.h (hunk 6 above)
Comment (fauxpark) [reply tmk_core/common/host_driver.h]: I think I want to move away from using `host_driver_t`, since even with the Digitizer feature disabled, the (null) function pointer will probably still take up a couple extra bytes. Of course this is also true for the mouse and extrakeys stuff, but they are quite often enabled so I don't see it as that much of an issue.

Comment (a-chol) [reply tmk_core/common/host_driver.h]: Agreed. Should I use a function pointer outside of the host_driver_t conditioned on DIGITIZER_ENABLE or do you have another suggestion.

File: keyboards/handwired/onekey/keymaps/digitizer/config.h (hunk 8 above)
Comment (a-chol) [reply keyboards/handwired/onekey/keymaps/digitizer/config.h:1]: Yeah that makes sense obviously, oversight on my part too.

**Comment by a-chol** [comment]
I saw the PR #12819 as well. Should I preemptively apply the fix for the digitizer report as well?

**Comment by a-chol** [comment]
Just to sum up the outstanding issues : 

- the linter doesn't seem to find the same formatting issues in the docker container running on my computer than the CI one. Is it a problem.
- the send_digitizer function pointer at the interface between tmk and the usb HW-dependent implementation adds a cost whether digitizer is enabled or not. Is Hiding it behind ifdef DIGITIZER_ENABLE while keeping it in the host_driver_t struct a good enough fix in the scope of this PR?
- Do I apply the fix from #12819 ?

**Comment by fauxpark** [comment]
> the linter doesn't seem to find the same formatting issues in the docker container running on my computer than the CI one. Is it a problem?

It's a problem in the sense that clang-format's behaviour isn't consistent across versions. The solution for now is to always use the Docker container when running a format pass, otherwise when the PR is merged qmk-bot will open a subsequent PR to format it "properly".

> the send_digitizer function pointer at the interface between tmk and the usb HW-dependent implementation adds a cost whether digitizer is enabled or not. Is Hiding it behind ifdef DIGITIZER_ENABLE while keeping it in the host_driver_t struct a good enough fix in the scope of this PR?

I'd prefer to have it implemented the way Joystick is, without touching the `host_driver_t` stuff at all.

> Do I apply the fix from #12819 ?

If that PR goes in first, yes. Otherwise, we'd ask for it to be rebased to apply the changes to the Digitizer stuff as well.

**Comment by a-chol** [comment]
So I guess this PR is ready to be merged, at least feature-wise. The CI failed on something that was not related to the test and is failing to launch now, is there a way to relaunch it when it's not blocked by docker anymore?

File: keyboards/handwired/onekey/keymaps/digitizer/keymap.c [hunk 9]
```diff
@@ -0,0 +1,26 @@
+#include QMK_KEYBOARD_H
```
Comment (drashna) [inline keyboards/handwired/onekey/keymaps/digitizer/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/handwired/onekey/keymaps/digitizer/keymap.c [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+#include "digitizer.h"
+
+#include "math.h"
+
+#ifndef ADC_PIN
```
Comment (drashna) [inline keyboards/handwired/onekey/keymaps/digitizer/keymap.c:7]: This should go into the config.h, not here, as this won't be picked up by anything but the keymap.c file.

File: quantum/digitizer.c [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "digitizer.h"
+
+digitizer_t digitizer = {.tipswitch = 0, .inrange = 0, .id = 0, .x = 0, .y = 0, .status = DZ_INITIALIZED | DZ_UPDATED};
+
+__attribute__((weak)) void digitizer_task(void) {
```
Comment (drashna) [inline quantum/digitizer.c:20]: Honestly, I'd kind of like this to be more like the pointing device code. 

Eg, a task and a send function, which handles the different aspects.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/handwired/onekey/keymaps/digitizer/keymap.c (hunk 10 above)
Comment (a-chol) [reply keyboards/handwired/onekey/keymaps/digitizer/keymap.c]: I just removed the ADC stuff as I didn't use it anyway.

File: docs/feature_digitizer.md [hunk 12]
```diff
@@ -0,0 +1,34 @@
+## Digitizer
```
Comment (drashna) [inline docs/feature_digitizer.md:1]: Could you add the functions used for the feature, as well as provide a simple example?

File: keyboards/handwired/onekey/keymaps/digitizer/keymap.c [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+#include "digitizer.h"
+
+#include "math.h"
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {LAYOUT_ortho_1x1(KC_A)};
+
+uint32_t timer = 0;
+
+void matrix_scan_user() {
```
Comment (drashna) [inline keyboards/handwired/onekey/keymaps/digitizer/keymap.c:26]:
```suggestion
void digitizer_task(void) {
```

File: keyboards/handwired/onekey/keymaps/digitizer/keymap.c [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        return;
+    }
+    
+    timer               = timer_read32();
+    digitizer_t digitizer;
+    digitizer.x         = 0.5 - 0.2 * cos(timer_read() / 250. / 6.28);
+    digitizer.y         = 0.5 - 0.2 * sin(timer_read() / 250. / 6.28);
+    digitizer.tipswitch = 0;
+    digitizer.inrange   = 1;
+    digitizer_set_report(digitizer);
```
Comment (drashna) [inline keyboards/handwired/onekey/keymaps/digitizer/keymap.c:37]: Shouldn't this have the send? 
```suggestion
    digitizer_set_report(digitizer);
    digitizer_send();
```

File: quantum/digitizer.c [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "digitizer.h"
+
+digitizer_t digitizerReport = {.tipswitch = 0, .inrange = 0, .id = 0, .x = 0, .y = 0, .status = DZ_INITIALIZED};
+
+__attribute__((weak)) void digitizer_send(void) {
+    if (digitizerReport.status & DZ_UPDATED) {
```
Comment (drashna) [inline quantum/digitizer.c:21]: Could prefer if this checks the actual status, eg compare the values, and see if it changed, rather than assuming that because the report set function was called that it has changed.

**Review by drashna — CHANGES_REQUESTED** [review]
Just a couple of things, otherwise, looks good

File: docs/feature_digitizer.md (hunk 12 above)
Comment (a-chol) [reply docs/feature_digitizer.md:1]: Yeah ,I forgot to update the documentation to match the new API. It'll be updated with the next commit.

File: keyboards/handwired/onekey/keymaps/digitizer/keymap.c (hunk 14 above)
Comment (a-chol) [reply keyboards/handwired/onekey/keymaps/digitizer/keymap.c:37]: pointing_device_task() does call pointing_device_send(); so it seems calling send explicitly is not striclty required. It makes sense to let the firmware chose when to actually send the report I guess, so I was aiming towards that API. Now this approach is intertwined with the previous review regarding using digitizer_task to update the report : either digitizer_task implements some internal behavior, then replacing it shouldn't be a solution for the end user. Or digitizer_task is public api, then it shouldn't implement internal behavior that the user could fail to reproduce and wonder why it's not working. One solution to this would be a firmware-level digitizer task and a keyboard/user-level digitizer_task, which ends up reimplementing the multi-level matrix_scan routine. 
Adding to that, the most common use case for digitizer will be sending them on keystrokes, so opening up the digitizer_task as a public API doesn't seem that useful.
Basically the two options are : 

- digitizer_task does nothing, the user has to send explicitly (which is how pointing device is advised to be used, but not how it's implemented)
- digitizer_task sends updates when one is detected, only the set_report function is useful.

File: quantum/digitizer.c (hunk 15 above)
Comment (a-chol) [reply quantum/digitizer.c:21]: Yeah, that's how pointing device works, but a (0,0) absolute position is a valid report to send consecutively, for instance if you want a key to reset the pointer position in the middle of the screen to find it more easily. Using an invalid value to denote a "not updated" state is a bit wacky and adds a lot of code compared to a flag which semantic is well defined. The flag should be out of the report though, to avoid exposing it publicly.

**Comment by a-chol** [comment]
I updated the changes requested, everything looks good now.

**Review by drashna — APPROVED** [review]

**Review by tzarc — APPROVED** [review]
